File: Data/Repositories/AuthorRepo.cs
﻿using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using Dapper;
using Dommel;
using EntityModels;
using Infrastructure.Data;
using Shared.Services;

namespace Data.Repositories
{
    public class AuthorRepo : BaseRepo, IAuthorRepo
    {
        public long Create(AuthorEM entity)
        {
            using (var con = Connection)
            {
                var result = con.Insert(entity);
                var converted = (long)Convert.ChangeType(result, typeof(long));
                return converted;
            }
        }

        public void Update(AuthorEM entity)
        {
            using (var con = Connection)
            {
                con.Update(entity);
            }
        }

        public void Delete(long id)
        {
            using (var con = Connection)
            {
                var SP = "USPAuthorDelete";
                var queryParameters = new DynamicParameters();
                queryParameters.Add("@Id", id);

                con.Query(SP, queryParameters, commandType: CommandType.StoredProcedure);
            }
        }

        public IEnumerable<AuthorEM> Get(string term)
        {
            using (var con = Connection)
            {
                var SP = "USPAuthorGetByTerm";
                var queryParameters = new DynamicParameters();
                queryParameters.Add("@Term", term);

                var authors = con.Query<AuthorEM>(SP, queryParameters, commandType: CommandType.StoredProcedure);

                return authors;
            }
        }

        public AuthorEM Get(long id)
        {
            using (var con = Connection)
            {
                var SP = "USPAuthorGet";
                var queryParameters = new DynamicParameters();
                queryParameters.Add("@Id", id);

                var result = con.QueryMultiple(SP, queryParameters, commandType: CommandType.StoredProcedure);

                var mappingObject = new AuthorMappingObjectEM();

                mappingObject.Author = result.ReadSingleOrDefault<AuthorEM>();
                mappingObject.Books = result.Read<BookEM>().AsList();

                var author = MapperService.Map<AuthorEM>(mappingObject);

                return author;
            }
        }

        public DataTableResponseEM<AuthorEM> GetList(DataTableRequestEM model)
        {
            using (var con = Connection)
            {
                var SPname = "USPAuthorGetList";

                var order = model.Order[0];
                var orderColumnName = model.Columns[order.Column].Name;

                var queryParameters = new DynamicParameters();
                queryParameters.Add("@Start", model.Start);
                queryParameters.Add("@Lenght", model.Length);
                queryParameters.Add("@OrderColumnName", orderColumnName);
                queryParameters.Add("@IsAsc", order.IsAcs);

                var reader = con.QueryMultiple(SPname, queryParameters, commandType: CommandType.StoredProcedure);
                var generalInfo = reader.ReadSingle<DataTableResponseGeneralEM>();

                var datatableResponse = new DataTableResponseEM<AuthorEM>();
                datatableResponse.RecordsFiltered = generalInfo.RecordsFiltered;
                datatableResponse.RecordsTotal = generalInfo.RecordsTotal;

                var authors = reader.Read<AuthorEM>();
                var authorBook = reader.Read<AuthorBookEM>();
                var books = reader.Read<BookEM>();

               
                datatableResponse.Data = MapWithBooks(authors, authorBook, books);

                return datatableResponse;
            }
        }

        public IEnumerable<AuthorEM> GetALL()
        {
            using (var con = Connection)
            {
                var authors = con.GetAll<AuthorEM>();

                var reader = con.QueryMultiple(@"SELECT AuthorID, BookID FROM AuthorBook
                                                 SELECT B.Id, B.[Name], B.[CreatedDate], B.Pages, B.Rate FROM Book B
                                                 WHERE EXISTS(SELECT 1 FROM AuthorBook AB WHERE AB.BookID = B.ID)");

                var authorBook = reader.Read<AuthorBookEM>();
                var books = reader.Read<BookEM>();

                return MapWithBooks(authors, authorBook, books);
            }
        }

        private IEnumerable<AuthorEM> MapWithBooks(IEnumerable<AuthorEM> authors, IEnumerable<AuthorBookEM> authorBook, IEnumerable<BookEM> books)
        {
            foreach (var a in authors)
            {
                var bookIds = authorBook.Where(ab => ab.AuthorId == a.Id.Value).Select(s => s.BookId).ToList();
                var booksToAdd = books.Where(b => bookIds.Contains(b.Id.Value));
                a.Books.AddRange(booksToAdd);
            }

            return authors;
        }
    }
}

File: DataVB/Repositories/AuthorRepo.vb
Imports Dapper
Imports Dommel
Imports EntityModels
Imports Infrastructure.Data
Imports [Shared].Services

Namespace Repositories

    Public Class AuthorRepo
        Inherits BaseRepo
        Implements IAuthorRepo

        Public Sub Update(entity As AuthorEM) Implements ICRUD(Of AuthorEM, Long).Update
            Using con = Connection
                con.Update(entity)
            End Using
        End Sub

        Public Sub Delete(id As Long) Implements ICRUD(Of AuthorEM, Long).Delete
            Using con = Connection
                Dim SP = "USPAuthorDelete"
                Dim queryParameters = New DynamicParameters()
                queryParameters.Add("@Id", id)

                con.Query(SP, queryParameters, commandType:=CommandType.StoredProcedure)
            End Using
        End Sub

        Public Function [Get](term As String) As IEnumerable(Of AuthorEM) Implements IAuthorRepo.Get
            Using con = Connection
                Dim SP = "USPAuthorGetByTerm"
                Dim queryParameters = New DynamicParameters()
                queryParameters.Add("@Term", term)

                Dim authors = con.Query(Of AuthorEM)(SP, queryParameters, commandType:=CommandType.StoredProcedure)

                Return authors
            End Using
        End Function

        Public Function [Get](id As Long) As AuthorEM Implements ICRUD(Of AuthorEM, Long).Get
            Using con = Connection
                Dim SP = "USPAuthorGet"
                Dim queryParameters = New DynamicParameters()
                queryParameters.Add("@Id", id)

                Dim result = con.QueryMultiple(SP, queryParameters, commandType:=CommandType.StoredProcedure)

                Dim mappingObject = New AuthorMappingObjectEM()

                mappingObject.Author = result.ReadSingleOrDefault(Of AuthorEM)()
                mappingObject.Books = result.Read(Of BookEM)().AsList()

                Dim author = MapperService.Map(Of AuthorEM)(mappingObject)

                Return author
            End Using
        End Function

        Public Function GetList(model As DataTableRequestEM) As DataTableResponseEM(Of AuthorEM) Implements IAuthorRepo.GetList
            Using con = Connection
                Dim SPname = "USPAuthorGetList"

                Dim order = model.Order(0)
                Dim orderColumnName = model.Columns(order.Column).Name

                Dim queryParameters = New DynamicParameters()
                queryParameters.Add("@Start", model.Start)
                queryParameters.Add("@Lenght", model.Length)
                queryParameters.Add("@OrderColumnName", orderColumnName)
                queryParameters.Add("@IsAsc", order.IsAcs)

                Dim reader = con.QueryMultiple(SPname, queryParameters, commandType:=CommandType.StoredProcedure)
                Dim generalInfo = reader.ReadSingle(Of DataTableResponseGeneralEM)()

                Dim datatableResponse = New DataTableResponseEM(Of AuthorEM)()
                datatableResponse.RecordsFiltered = generalInfo.RecordsFiltered
                datatableResponse.RecordsTotal = generalInfo.RecordsTotal

                Dim authors = reader.Read(Of AuthorEM)().AsList()

                Dim authorBook = reader.Read(Of AuthorBookEM)().AsList()

                Dim books = reader.Read(Of BookEM)().AsList()

                For Each a In authors
                    Dim bookIds = authorBook.Where(Function(ab) ab.AuthorId = a.Id.Value).Select(Function(s) s.BookId).ToList()
                    Dim booksToAdd = books.Where(Function(b) bookIds.Contains(b.Id.Value))
                    a.Books.AddRange(booksToAdd)
                Next

                datatableResponse.Data = authors

                Return datatableResponse
            End Using
        End Function

        Public Function Create(entity As AuthorEM) As Long Implements ICRUD(Of AuthorEM, Long).Create
            Using con = Connection
                Dim result = con.Insert(entity)
                Dim converted = CLng(result)
                Return converted
            End Using
        End Function
    End Class

End Namespace
